homeassistant/components/mqtt/config_flow.py

```diff
@@ -721,6 +859,304 @@ def _validate(
         )
 
 
+class MQTTSubentryFlowHandler(ConfigSubentryFlow):
+    """Handle MQTT subentry flow."""
+
+    _subentry_data: MqttSubentryData
+    _object_id: str | None = None
+    _component_id: str | None = None
+    _dirty: bool = False
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Add a subentry."""
+        self._subentry_data = MqttSubentryData(device=MqttDeviceData(), components={})
+        return await self.async_step_device()
+
+    async def async_step_reconfigure(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Reconfigure a subentry."""
+        reconfigure_subentry = self._get_reconfigure_subentry()
+        self._subentry_data = cast(
+            MqttSubentryData, deepcopy(dict(reconfigure_subentry.data))
+        )
+        return await self.async_step_summary_menu()
+
+    async def async_step_device(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Add a new MQTT device."""
+        errors: dict[str, str] = {}
+        validate_field("configuration_url", cv.url, user_input, errors, "invalid_url")
+        if not errors and user_input is not None:
+            self._subentry_data[CONF_DEVICE] = cast(MqttDeviceData, user_input)
+            if self.source == SOURCE_RECONFIGURE:
+                self._dirty = True
+                return await self.async_step_summary_menu()
+            return await self.async_step_entity()
+
+        data_schema = MQTT_DEVICE_SCHEMA
+        data_schema = self.add_suggested_values_to_schema(
+            data_schema,
+            self._subentry_data[CONF_DEVICE] if user_input is None else user_input,
+        )
+        return self.async_show_form(
+            step_id=CONF_DEVICE,
+            data_schema=data_schema,
+            errors=errors,
+            last_step=False,
+        )
+
+    async def async_step_entity(
+        self, user_input: dict[str, Any] | None = None
+    ) -> SubentryFlowResult:
+        """Add or edit an mqtt entity."""
+        errors: dict[str, str] = {}
+        device_name = self._subentry_data[CONF_DEVICE][CONF_NAME]
+        data_schema_fields = COMMON_ENTITY_FIELDS
+        data_schema = data_schema_from_fields(data_schema_fields)
+        if user_input is not None:
+            # Add the component but check if the object ID is unique and not reconfiguring
+            validate_user_input(user_input, data_schema_fields, errors)
+            platform_object_id = (
+                f"{user_input['platform']}_{slugify(user_input.get('name', 'none'))}"
+            )
+            if self._object_id != platform_object_id:
+                if platform_object_id in self._subentry_data["components"]:
+                    errors[CONF_NAME] = "name_not_unique"
```

Comment:
This is a weird limitation, and breaks down for languages which use accents:
```py
>>> from homeassistant.util import slugify
>>> slugify("Höna")
'hona'
>>> slugify("Hona")
'hona'
>>>
```

Can we just generate a UUID for each added entity and use that, instead of deriving an object id from the name?